webapi/src/main/scala/org/knora/webapi/responders/v1/UsersResponderV1.scala

```diff
@@ -20,144 +20,369 @@
-            }
+            //_ = log.debug(MessageUtil.toSource(userDataQueryResponse))
 
-            userProfileV1 = {
-                if (groupedUserData.isEmpty) {
-                    None
-                } else {
-                    Some(UserProfileV1(userDataV1, groupIris, projectIris))
-                }
+            _ = if (userDataQueryResponse.results.bindings.isEmpty) {
+                throw NotFoundException(s"User '$userIri' not found")
             }
 
+            userProfileV1 = userDataQueryResponse2UserProfile(userDataQueryResponse, clean)
 
-        } yield userProfileV1
+        } yield userProfileV1 // UserProfileV1(userDataV1, groupIris, projectIris)
     }
 
     /**
       * Gets information about a Knora user, and returns it in a [[Option[UserProfileV1]].
+      *
       * @param username the username of the user.
       * @return a [[Option[UserProfileV1]] describing the user.
       */
-    private def getUserProfileByUsernameV1(username: String): Future[Option[UserProfileV1]] = {
+    private def getUserProfileByUsernameV1(username: String, clean: Boolean): Future[UserProfileV1] = {
+        log.debug(s"getUserProfileByUsernameV1('$username', '$clean') called")
         for {
             sparqlQuery <- Future(queries.sparql.v1.txt.getUserByUsername(username).toString())
             userDataQueryResponse <- (storeManager ? SparqlSelectRequest(sparqlQuery)).mapTo[SparqlSelectResponse]
 
-            //_ = println(MessageUtil.toSource(userDataQueryResponse))
+            //_ = log.debug(MessageUtil.toSource(userDataQueryResponse))
 
             _ = if (userDataQueryResponse.results.bindings.isEmpty) {
-                throw NotFoundException(s"User $username not found")
+                throw NotFoundException(s"User '$username' not found")
             }
 
-            userIri = userDataQueryResponse.getFirstRow.rowMap("s")
+            userProfileV1 = userDataQueryResponse2UserProfile(userDataQueryResponse, clean)
+
+        } yield userProfileV1 // UserProfileV1(userDataV1, groupIris, projectIris)
+    }
+
+    /**
+      * Creates a new user. Self-registration is allowed, so even the default user, i.e. with no credentials supplied,
+      * is allowed to create the new user.
+      *
+      * Referenced Websites:
+      *                     - https://crackstation.net/hashing-security.htm
+      *                     - http://blog.ircmaxell.com/2012/12/seven-ways-to-screw-up-bcrypt.html
+      *
+      * @param newUserData a [[NewUserDataV1]] object containing information about the new user to be created.
+      * @param userProfile a [[UserProfileV1]] object containing information about the requesting user.
+      * @return a future containing the [[UserOperationResponseV1]].
+      */
+    private def createNewUserV1(newUserData: NewUserDataV1, userProfile: UserProfileV1, apiRequestID: UUID): Future[UserOperationResponseV1] = {
+        for {
+            a <- Future("")
```

Comment:
I guess `a <- Future("")` is just here because you need a `<-` on the first line of the `for` comprehension? It's a bit ugly like this. Instead you could write:

```
// check if username or password are not empty
_ <- Future {
    if (newUserData.username.isEmpty) throw BadRequestException("Username cannot be empty")
    if (newUserData.password.isEmpty) throw BadRequestException("Password cannot be empty")
}

// check if the supplied username for the new user is unique, i.e. not already registered
sparqlQuery <- queries.sparql.v1.txt.getUserByUsername(newUserData.username).toString()
```